cluster/channel.go

```diff
@@ -0,0 +1,130 @@
+type Channel struct {
+	key          string
+	send         func([]byte)
+	peers        func() []*memberlist.Node
+	sendOversize func(*memberlist.Node, []byte) error
+
+	msgc   chan []byte
+	logger log.Logger
+
+	oversizeGossipMessageFailureTotal prometheus.Counter
+	oversizeGossipDuration            prometheus.Histogram
+}
+
+// NewChannel creates a new Channel struct, which handles sending normal and
+// oversize messages to peers.
+func NewChannel(
+	key string,
+	send func([]byte),
+	peers func() []*memberlist.Node,
+	sendOversize func(*memberlist.Node, []byte) error,
+	logger log.Logger,
+	stopc chan struct{},
+	reg prometheus.Registerer,
+) *Channel {
+	oversizeGossipMessageFailureTotal := prometheus.NewCounter(prometheus.CounterOpts{
+		Name:        "alertmanager_oversized_gossip_message_failure_total",
+		Help:        "Number notification log received queries that failed.",
+		ConstLabels: prometheus.Labels{"key": key},
+	})
+	oversizeGossipDuration := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Name:        "alertmanager_oversize_gossip_message_duration_seconds",
+		Help:        "Duration of oversized gossip message requests.",
+		ConstLabels: prometheus.Labels{"key": key},
+	})
+
+	reg.MustRegister(oversizeGossipDuration, oversizeGossipMessageFailureTotal)
+
+	c := &Channel{
+		key:                               key,
+		send:                              send,
+		peers:                             peers,
+		logger:                            logger,
+		msgc:                              make(chan []byte, 200),
+		sendOversize:                      sendOversize,
+		oversizeGossipMessageFailureTotal: oversizeGossipMessageFailureTotal,
+		oversizeGossipDuration:            oversizeGossipDuration,
+	}
+
+	go c.handleOverSizedMessages(stopc)
+
+	return c
+}
+
+// handleOverSizedMessages prevents memberlist from opening too many parallel
+// TCP connections to its peers.
+func (c *Channel) handleOverSizedMessages(stopc chan struct{}) {
+	var wg sync.WaitGroup
+	for {
+		select {
+		case b := <-c.msgc:
+			for _, n := range c.peers() {
+				wg.Add(1)
+				go func(n *memberlist.Node) {
+					defer wg.Done()
+
+					start := time.Now()
+					if err := c.sendOversize(n, b); err != nil {
+						level.Debug(c.logger).Log("msg", "failed to send reliable", "key", c.key, "node", n, "err", err)
+						c.oversizeGossipMessageFailureTotal.Inc()
+						return
+					}
+					c.oversizeGossipDuration.Observe(time.Since(start).Seconds())
+				}(n)
+			}
+
+			wg.Wait()
+		case <-stopc:
+			return
+		}
+	}
+}
+
+// Broadcast enqueues a message for broadcasting.
+func (c *Channel) Broadcast(b []byte) {
+	b, err := proto.Marshal(&clusterpb.Part{Key: c.key, Data: b})
+	if err != nil {
+		return
+	}
+
+	if OversizedMessage(b) {
+		c.msgc <- b
+	} else {
+		c.send(b)
+	}
+}
+
+// OversizedMessage indicates whether or not the byte payload should be sent
+// via TCP.
+func OversizedMessage(b []byte) bool {
+	return len(b) > maxGossipPacketSize/2
```

Comment:
More or less. From the data I was logging, it seemed like the overhead from memberlist was always under 5 bytes, which then leaves us with ~1395 bytes for the gossip message. I looked at the message sizes we're gossiping at SC, and the maximum in the last 2 days is 500 bytes, and average around 40 bytes.

We could definitely go higher, like 80 or 90% of the max. I didn't worry too much about it since, from the data I've seen here, messages are usually much smaller 800 bytes, or much larger than 800 bytes.

Reply:
Sounds good to me, I was just wondering where this came from.